File: frontend/src/types/notification.ts
export type NotificationType =
  | "pipeline"
  | "quota"
  | "collaboration"
  | "billing"
  | "security"
  | "system";

export interface NotificationItem {
  id: number;
  user_id: number;
  type: string;
  title: string;
  message: string;
  action_url: string | null;
  target_type: string | null;
  target_id: number | null;
  metadata: Record<string, any> | null;
  is_read: boolean;
  read_at: string | null;
  created_at: string;
}

export interface NotificationListResponse {
  items: NotificationItem[];
  total: number;
  unread_count: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface UnreadCountResponse {
  unread_count: number;
}

export interface MarkAllReadResponse {
  success: boolean;
  updated_count: number;
}

export interface NotificationPreferences {
  // Email Channels
  email_on_pipeline_success: boolean;
  email_on_pipeline_failed: boolean;
  email_on_quota_warning: boolean;
  email_on_project_invitation: boolean;
  email_on_comment_mention: boolean;

  // In-App Channels
  inapp_on_pipeline_success: boolean;
  inapp_on_pipeline_failed: boolean;
  inapp_on_quota_warning: boolean;
  inapp_on_project_invitation: boolean;
  inapp_on_comment_mention: boolean;
}

export type NotificationPreferencesPatch = Partial<NotificationPreferences>;

export interface TestAlertRequest {
  type?: string;
  title?: string;
  message?: string;
  action_url?: string;
}

export interface NotificationQueryParams {
  page?: number;
  page_size?: number;
  unread_only?: boolean;
  type?: string;
}

File: backend/app/schemas/notification.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


# =========================================================
# Request Schemas
# =========================================================

class NotificationPreferencesPatch(BaseModel):
    # Email Channels
    email_on_pipeline_success: Optional[bool] = None
    email_on_pipeline_failed: Optional[bool] = None
    email_on_quota_warning: Optional[bool] = None
    email_on_project_invitation: Optional[bool] = None
    email_on_comment_mention: Optional[bool] = None

    # In-App Channels
    inapp_on_pipeline_success: Optional[bool] = None
    inapp_on_pipeline_failed: Optional[bool] = None
    inapp_on_quota_warning: Optional[bool] = None
    inapp_on_project_invitation: Optional[bool] = None
    inapp_on_comment_mention: Optional[bool] = None


class TestAlertRequest(BaseModel):
    type: Optional[str] = Field(default="system", max_length=50)
    title: Optional[str] = Field(default=None, max_length=255)
    message: Optional[str] = Field(default=None, max_length=1000)
    action_url: Optional[str] = Field(default=None, max_length=500)


# =========================================================
# Response Schemas
# =========================================================

class NotificationResponse(BaseModel):
    id: int
    user_id: int
    type: str
    title: str
    message: str
    action_url: Optional[str] = None
    target_type: Optional[str] = None
    target_id: Optional[int] = None
    metadata: Optional[Dict[str, Any]] = None
    is_read: bool
    read_at: Optional[datetime] = None
    created_at: datetime


class NotificationListResponse(BaseModel):
    items: List[NotificationResponse]
    total: int
    unread_count: int
    page: int
    page_size: int
    total_pages: int


class UnreadCountResponse(BaseModel):
    unread_count: int


class MarkAllReadResponse(BaseModel):
    success: bool
    updated_count: int


class NotificationPreferencesResponse(BaseModel):
    email_on_pipeline_success: bool
    email_on_pipeline_failed: bool
    email_on_quota_warning: bool
    email_on_project_invitation: bool
    email_on_comment_mention: bool
    inapp_on_pipeline_success: bool
    inapp_on_pipeline_failed: bool
    inapp_on_quota_warning: bool
    inapp_on_project_invitation: bool
    inapp_on_comment_mention: bool
